<a href="https://colab.research.google.com/github/AFBA1993/Yandex/blob/main/TASK1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Candidate: Andrés Felipe Bolaños Acosta.


Position: code reviewer.


E-mail: afb.acosta@gmail.com.

Mobile phone: +5511973662765.

Portfolio: https://github.com/AFBA1993/.

Linkedin: https://www.linkedin.com/in/andr%C3%A9s-felipe-bola%C3%B1os-acosta-b14b691a4/.

Google Scholar: https://scholar.google.com.br/citations?user=GLKkmsoAAAAJ&hl=pt-BR.

### ***Task 1. Working with data***

In [1]:
import pandas as pd
import math
import numpy as np
from re import split

1.1. Download the data set movie_metadata.csv, which contains data about films from IMDb
(Internet Movie Database).

*Solution:*

In [2]:
path_task1="/content/drive/MyDrive/YandexTest/movie_metadata.csv"
movie_data=pd.read_csv(path_task1)

1.2. The duration column contains data on the film length. How many missing values are there
in this column? 

*Answer:*

In [3]:
movie_data['duration'].isnull().sum()

15

There are 15 missing values in the duration column.

1.3. Replace the missing values in the duration column with the median value for this column.

*Answer:*

In [4]:
movie_data.duration.fillna(movie_data.duration.median(),inplace = True)

1.4. What is the average film length? Give the answer as a floating-point figure rounded to
two decimal places.

*Answer:*

In [5]:
round(movie_data.duration.mean(),2)

107.19

The average film length is 107.19

1.5. Create a movie_duration_category column, which will contain three categories
depending on the film length:


*   Category "1. <90" if the film is less than 90 minutes long
*   Category "2. 90–120" if the film is between 90 minutes and two hours long (inclusively)
*   Category "3. >120" if the film is more than two hours long.

*Answer:*

In [6]:
def category_length(duration_each_film):
  if duration_each_film < 90:
    return "<90"
  elif duration_each_film >= 90 and duration_each_film <=120:
    return "90-120"
  elif duration_each_film > 120:
    return ">120"

In [7]:
movie_data["movie_duration_category"] = movie_data["duration"].apply(lambda 
                                                  film: category_length(film))

1.6. Build a summary table for films released after 2000 (inclusively), to list the numbers of
films:

*   Table rows: year
*   Table columns: movie duration category ("<90", "90–120", ">120")
*   The year of release should be displayed in the YYYY format.

*Answer:*

In [8]:
movies_after_2000 = movie_data.loc[(movie_data["title_year"]) > 1999]

In [9]:
Table_summary_movies = pd.DataFrame()

In [10]:
Table_summary_movies["year"] = movies_after_2000["title_year"]
Table_summary_movies["movie_title"] = movies_after_2000["movie_title"]
Table_summary_movies["movie_duration_category"] = movies_after_2000[
                                                    "movie_duration_category"]

In [11]:
Table_summary_movies["year"] = Table_summary_movies["year"].apply(
                                                        lambda year: int(year))

In [12]:
Table_summary_movies

year                                movie_title movie_duration_category
0     2009                                    Avatar                     >120
1     2007  Pirates of the Caribbean: At World's End                     >120
2     2015                                   Spectre                     >120
3     2012                     The Dark Knight Rises                     >120
5     2012                               John Carter                     >120
...    ...                                        ...                     ...
5037  2011                                 Newlyweds                   90-120
5038  2013                   Signed Sealed Delivered                      <90
5040  2013                      A Plague So Pleasant                      <90
5041  2012                          Shanghai Calling                   90-120
5042  2004                         My Date with Drew                   90-120

[3597 rows x 3 columns]

In [13]:
Table_summary_movies[">120"] = Table_summary_movies["movie_duration_category"]\
                         .apply(lambda category: 1 if category == ">120" else 0)

Table_summary_movies["<90"] = Table_summary_movies["movie_duration_category"]\
                         .apply(lambda category: 1 if category == "<90" else 0
                                )
Table_summary_movies["90-120"] = \
                       Table_summary_movies["movie_duration_category"]\
                      .apply(lambda category: 1 if category == "90-120" else 0)

Table_summary_movies

year                                movie_title movie_duration_category  \
0     2009                                    Avatar                     >120   
1     2007  Pirates of the Caribbean: At World's End                     >120   
2     2015                                   Spectre                     >120   
3     2012                     The Dark Knight Rises                     >120   
5     2012                               John Carter                     >120   
...    ...                                        ...                     ...   
5037  2011                                 Newlyweds                   90-120   
5038  2013                   Signed Sealed Delivered                      <90   
5040  2013                      A Plague So Pleasant                      <90   
5041  2012                          Shanghai Calling                   90-120   
5042  2004                         My Date with Drew                   90-120   

      >120  <90  90-120  
0        1    0       0  
1        1    0       0  
2        1    0       0  
3        1    0       0  
5        1    0       0  
...    ...  ...     ...  
5037     0    0       1  
5038     0    1       0  
5040     0    1       0  
5041     0    0       1  
5042     0    0       1  

[3597 rows x 6 columns]

In [14]:
Table_summary_movies.groupby(["year"])["<90", "90-120", ">120"].sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


<90  90-120  >120
year                   
2000   25     112    34
2001   29     120    39
2002   36     146    27
2003   31     108    30
2004   30     142    42
2005   31     142    48
2006   40     146    53
2007   31     130    43
2008   29     160    36
2009   42     178    40
2010   26     171    33
2011   36     150    39
2012   41     139    41
2013   38     145    54
2014   36     171    45
2015   34     145    47
2016    8      79    19

1.7. How many films between 90 minutes and two hours long were released in 2008?.

*Answer:*

In [15]:
movies_2008 = Table_summary_movies.loc[(Table_summary_movies["year"] == 2008 ) 
                    & (Table_summary_movies["movie_duration_category"] == 
                                                                    "90-120")]

In [16]:
movies_2008.reset_index(drop=True, inplace=True)

In [17]:
movies_2008.index[-1]+1

160

160 films were released in 2008

1.8. The plot_keywords column holds keywords characterizing the film's plot. Using the data
in this column, create a column called movie_plot_category, to contain four categories
depending on the key words in the column:



*   Category "love_and_death" if the keywords include both "love" and "death".
*  Category "love" if the keywords include the word "love".
*   Category "death" if the keywords include the word "death".
*   Category "other" if the keywords do not meet the conditions above.

*Answer:*

In [18]:
def plot_category(keywords_text):
  def isnan(value):
    try:
        return math.isnan(float(value))
    except:
        return False

  if isnan(keywords_text):
    return np.nan
  else: 
    just_love = ("love" in keywords_text) and not ("death" in keywords_text)
    just_death = not("love" in keywords_text) and ("death" in keywords_text)
    love_death = ("love" in keywords_text) and ("death" in keywords_text)

    if just_love :
      return "love"  
    elif just_death:
      return "death"
    elif love_death:
      return "love_and_death"
    else:
      return "other"  

def plot_category2(keywords_text):
  def isnan(value):
    try:
        import math
        return math.isnan(float(value))
    except:
        return False

  if isnan(keywords_text):
    return np.nan
  else:

    love = False
    death = False

    keywords_text= keywords_text.split("|")
    for word in keywords_text:
      print(word)
      word = word.split()
      for subword in word:
        print("  "+subword)
        if subword == "love":
          love = True
        if subword == "death":
          death = True

    if love and not death:
      return "love"
    elif death and not love:
      return "death"
    elif death and love:
      return "love_and_death"
    else:
      return "other",love,death

In [19]:
movie_data["movie_plot_category"] = movie_data["plot_keywords"].apply(lambda 
                                                  film: plot_category(film))

*Note:* The current solution uses the function "plot_category" in order to clasify the films, this function carries out the classification if the especified word is within the "plot_keywords". 

However, a second functions was coded, which is named as "plot_category2", this function classifies the film if the specified category word matches exactly in the "plot_keywords" field.

1.9. The imdb_score column shows a viewer rating for the film. Build a table to reflect the average rating of films depending on which movie_plot_category category they belong to.

*Answer:*

In [20]:
movie_data["imdb_score"].isnull().sum()

0

In [21]:
other_cat_films = movie_data.loc[(movie_data["movie_plot_category"] == "other")]
love_cat_films = movie_data.loc[(movie_data["movie_plot_category"] == "love")]
death_cat_films = movie_data.loc[(movie_data["movie_plot_category"] == "death")]
lovedeath_cat_films = movie_data.loc[(movie_data["movie_plot_category"]
                                                                   == "death")]

In [22]:
avg_other = other_cat_films["imdb_score"].mean()
avg_love =  love_cat_films["imdb_score"].mean()
avg_death = death_cat_films["imdb_score"].mean()
avg_love_death = death_cat_films["imdb_score"].mean()


In [23]:
category_avg_score = {"category" : ["love", "death", "love_and_death", "other"],
                      "averaged_imdb_score" : [avg_love, avg_death, 
                                               avg_love_death, avg_other]}

In [24]:
category_avg_score = pd.DataFrame(category_avg_score)
category_avg_score

category  averaged_imdb_score
0            love             6.580769
1           death             6.535465
2  love_and_death             6.535465
3           other             6.443017

In [25]:
movie_data.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes',
       'movie_duration_category', 'movie_plot_category'],
      dtype='object')

1.10. What is the average rating of films in the "love" category? Give the answer as a floatingpoint figure rounded to two decimal places:

*Answer:*

In [26]:
love_movies = movie_data.loc[(movie_data["movie_plot_category"] == "love")]

In [27]:
love_movies["imdb_score"].isnull().sum()

0

In [28]:
round(love_movies["imdb_score"].mean(),2)

6.58

The average rating of movies in the love category is 6.58 the filtered values related to such a category does not contain null values.

1.11. The budget column contains the film's budget. What is the median budget for all the films
listed? Give the answer as an integer.

*Answer:*

In [29]:
movie_data["budget"] = movie_data["budget"].apply(lambda 
                                            film: float(film.replace("$","")))

In [30]:
int(movie_data["budget"].median())

15000000

The median of movies' budget is 1,500,000.00 US Dollars. 
